lib/vrl/DESIGN.md

```diff
@@ -0,0 +1,269 @@
+  needs.
+
+- user-defined functions
+
+  User-defined functions again produce indirection. While it _might_ be useful
+  to some extremely large use-cases, in most cases, allowing people to read
+  a program from top to bottom without having to jump around is more clear in
+  the context within which VRL is used (remapping).
+
+- network calls (see [#4517][] and [#8717][])
+
+  In order to avoid performance footguns, we want to ensure each function is as
+  performant as it can be, and there's no way to use functions in such a way
+  that performance of a VRL program tanks. We might introduce network calls at
+  some point, if we find a good caching solution to solve most of our concerns,
+  but so far we've avoided any network calls inside our stdlib.
+
+[#5507]: https://github.com/timberio/vector/issues/5507
+[#4517]: https://github.com/timberio/vector/issues/4517#issuecomment-754160338
+[#8717]: https://github.com/timberio/vector/pull/8717
+
+## Conventions
+
+### Syntax
+
+Keep VRL as syntax-light as possible. The less symbols, the more readable a VRL
+program is.
+
+Use functions whenever possible, and only introduce new syntax if a common
+pattern warrants a more convenient syntax-based solution, or functions are too
+limited in their capability to solve the required need.
+
+### Functions
+
+#### Composition
+
+- Favor function composition over single-purpose functions.
+
+- If a problem needs to be solved in multiple steps, consider adding
+  single-purpose functions for each individual step.
+
+- If useability or readability is hurt by composition, favor single-purpose
+  functions.
+
+#### Naming
+
+- Function names are lower-cased (e.g. `round`).
+
+- Multi-name functions use underscore (`_`) as a separator (e.g.
+  `encode_key_value`).
+
+- Favor explicit verbose names over terse ones (e.g.
+  `parse_aws_cloudwatch_log_subscription_message` over `parse_aws_cwl_msg`).
+
+- Functions should be preceded with their function category for organization and
+  discovery.
+
+  - Use `parse_*` for string to type decoding functions (e.g. `parse_json` and
+    `parse_grok`).
+  
+  - Use `decode_*` for string to string decoding functions (e.g.
+    `decode_base64`).
+  
+  - Use `encode_*` for string encoding functions (e.g. `encode_base64`).
+  
+  - Use `to_*` to convert from one type to another (e.g. `to_bool`).
+
+  - Use `is_*` to determine if the provided value is of a given type (e.g.
+    `is_string` or `is_json`).
+  
+  - Use `format_*` for string formatting functions (e.g. `format_timestamp` and
+    `format_number`).
+
+#### Return Types
+
+- Return boolean from `is_*` functions (e.g. `is_string`).
+
+- Return a string from `encode_*` functions (e.g. `encode_base64`).
+
+- Return an error whenever the function can fail at runtime.
+
+#### Mutability
+
+As a general rule, functions never mutate values.
+
+Favor this:
+
+```coffee
+# explicitly assign the parsed JSON to `.message`
+.message = parse_json(.message)
```

Comment:
Do we want to keep the examples valid VRL?

Reply:
It's a snippet of a valid VRL program, but I don't think we need to show the entire program to explain a particular pattern, especially given that this is focussed on developers, not users.